File: android/app/src/main/kotlin/com/siezagym/app/Features/Routines/RoutineDetailScreen.kt
package com.siezagym.app.Features.Routines

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ExpandMore
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.siezagym.app.DesignSystem.CreditoGifs
import com.siezagym.app.DesignSystem.GlassCard
import com.siezagym.app.DesignSystem.LocalD2Theme
import com.siezagym.app.DesignSystem.Miniatura
import com.siezagym.app.DesignSystem.PanelLista
import com.siezagym.app.DesignSystem.Pantalla
import com.siezagym.app.DesignSystem.SectionLabel
import com.siezagym.app.DesignSystem.SolidButton
import com.siezagym.app.DesignSystem.StatsCard
import com.siezagym.app.DesignSystem.Vacio
import com.siezagym.app.DesignSystem.WidgetMeter
import com.siezagym.app.Domain.RoutineSummary
import com.siezagym.app.Features.Shared.NavInset
import com.siezagym.app.Models.Exercise
import com.siezagym.app.Models.PlannedSet
import com.siezagym.app.Models.RegistrationType
import com.siezagym.app.Models.Routine
import com.siezagym.app.Models.RoutineExercise
import com.siezagym.app.Services.GymStore
import androidx.compose.runtime.collectAsState
import java.util.Locale
import kotlin.math.roundToInt

/** El detalle de una rutina, igual que `/rutinas/<id>` en la web: los tres
 *  números, la lista de ejercicios que se despliega de a uno, el reparto
 *  muscular, y abajo el botón de empezar. */
@Composable
fun RoutineDetailScreen(
    routine: Routine,
    store: GymStore,
    volver: () -> Unit,
    onEmpezar: (Routine) -> Unit,
) {
    val tema = LocalD2Theme.current
    val catalogo by store.catalog.collectAsState()

    var abierto by remember { mutableStateOf<String?>(null) }

    val minutos = RoutineSummary.estimatedMinutes(routine, catalogo)
    val reparto = RoutineSummary.muscleDistribution(routine, catalogo)
    val hayGifs = routine.exercises.any { catalogo[it.exerciseID]?.mediaUrl != null }

    Box(Modifier.fillMaxSize()) {
        Pantalla(
            titulo = routine.name,
            rotulo = if (routine.isAssigned) "Rutina del coach" else null,
            volver = true,
            onVolver = volver,
        ) {
            Column {
                StatsCard(
                    datos = listOf(
                        "${routine.exercises.count()}" to if (routine.exercises.count() == 1) "ejercicio" else "ejercicios",
                        "${routine.totalSets}" to if (routine.totalSets == 1) "serie" else "series",
                        // Es una cuenta sobre las series prescritas, no un tiempo
                        // medido: la pantalla lo dice.
                        "$minutos" to "min estimados",
                    ),
                    modifier = Modifier.padding(top = 20.dp),
                )

                SectionLabel(
                    "Ejercicios · ${routine.exercises.count()}",
                    modifier = Modifier.padding(top = 24.dp, bottom = 10.dp),
                )

                if (routine.exercises.isEmpty()) {
                    Vacio(texto = "Esta rutina no tiene ejercicios.")
                } else {
                    PanelLista {
                        routine.exercises.forEachIndexed { indice, item ->
                            if (indice > 0) {
                                Box(Modifier.fillMaxWidth().height(1.dp).background(tema.borde))
                            }
                            val clave = "$indice-${item.exerciseID}"
                            FilaEjercicio(
                                item = item,
                                ejercicio = catalogo[item.exerciseID],
                                nombre = store.name(item.exerciseID),
                                abierto = abierto == clave,
                                alTocar = {
                                    abierto = if (abierto == clave) null else clave
                                },
                            )
                        }
                    }
                }

                if (reparto.isNotEmpty()) {
                    SectionLabel(
                        "Músculos que trabaja",
                        modifier = Modifier.padding(top = 24.dp, bottom = 10.dp),
                    )

                    GlassCard {
                        Column(verticalArrangement = Arrangement.spacedBy(11.dp)) {
                            reparto.take(5).forEach { fila ->
                                Column(verticalArrangement = Arrangement.spacedBy(5.dp)) {
                                    Row(verticalAlignment = Alignment.CenterVertically) {
                                        Text(fila.muscle.label, fontSize = 12.sp, color = tema.texto)
                                        Spacer(Modifier.weight(1f))
                                        Text(
                                            "${(fila.pct * 100).roundToInt()}%",
                                            fontSize = 12.sp,
                                            color = tema.texto2,
                                        )
                                    }
                                    WidgetMeter(proporcion = fila.pct.toFloat())
                                }
                            }
                        }
                    }
                }

                if (hayGifs) {
                    CreditoGifs()
                }

                // Aire para que el último ejercicio no quede abajo del botón.
                Spacer(Modifier.height(90.dp))
            }
        }

        SolidButton(
            texto = "Comenzar entrenamiento",
            onClick = { onEmpezar(routine) },
            enabled = routine.exercises.isNotEmpty(),
            modifier = Modifier
                .align(Alignment.BottomCenter)
                .padding(horizontal = 18.dp)
                .padding(bottom = 12.dp + NavInset.bottom.dp),
        )
    }
}

/** Un ejercicio de la rutina. Cerrado muestra el resumen de lo prescrito;
 *  abierto, una fila por serie. */
@Composable
private fun FilaEjercicio(
    item: RoutineExercise,
    ejercicio: Exercise?,
    nombre: String,
    abierto: Boolean,
    alTocar: () -> Unit,
) {
    val tema = LocalD2Theme.current
    val series = series(item)
    val esDeTiempo = ejercicio?.registrationType?.isTimeBased == true
    val llevaPeso = ejercicio?.registrationType == RegistrationType.PESO_REPS
    val muestraRIR = series.any { it.rir != null }

    Column {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            modifier = Modifier
                .fillMaxWidth()
                .clickable(onClick = alTocar)
                .padding(horizontal = 14.dp, vertical = 12.dp),
        ) {
            Miniatura(url = ejercicio?.mediaUrl, lado = 54.dp)
            Spacer(Modifier.width(12.dp))
            Column(Modifier.weight(1f)) {
                Text(
                    nombre,
                    fontSize = 14.sp,
                    fontWeight = FontWeight.Medium,
                    color = tema.texto,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
                Spacer(Modifier.height(4.dp))
                Text(ejercicio?.primaryMuscle?.label ?: "Sin datos", fontSize = 11.sp, color = tema.texto2)
            }
            Spacer(Modifier.width(12.dp))
            Text(resumen(series, esDeTiempo), fontSize = 12.sp, color = tema.texto2, maxLines = 1)
            Spacer(Modifier.width(8.dp))
            val rotacion by animateFloatAsState(if (abierto) 180f else 0f, label = "chevron")
            Icon(
                Icons.Filled.ExpandMore,
                contentDescription = null,
                tint = tema.texto3,
                modifier = Modifier
                    .size(14.dp)
                    .graphicsLayer { rotationZ = rotacion },
            )
        }

        if (abierto) {
            Column(
                verticalArrangement = Arrangement.spacedBy(6.dp),
                modifier = Modifier
                    .padding(horizontal = 14.dp)
                    .padding(bottom = 14.dp),
            ) {
                Row(verticalAlignment = Alignment.CenterVertically, modifier = Modifier.fillMaxWidth()) {
                    Box(Modifier.width(22.dp))
                    Text(
                        if (esDeTiempo) "Tiempo" else "Reps",
                        fontSize = 10.sp,
                        color = tema.texto3,
                        modifier = Modifier.weight(1f),
                        textAlign = TextAlign.Center,
                    )
                    if (llevaPeso) {
                        Text(
                            "Peso",
                            fontSize = 10.sp,
                            color = tema.texto3,
                            modifier = Modifier.weight(1f),
                            textAlign = TextAlign.Center,
                        )
                    }
                    if (muestraRIR) {
                        Text(
                            "RIR",
                            fontSize = 10.sp,
                            color = tema.texto3,
                            modifier = Modifier.weight(1f),
                            textAlign = TextAlign.Center,
                        )
                    }
                }

                series.forEach { serie ->
                    Row(verticalAlignment = Alignment.CenterVertically, modifier = Modifier.fillMaxWidth()) {
                        Text(
                            "${serie.setNumber}",
                            fontSize = 11.sp,
                            color = tema.texto3,
                            modifier = Modifier.width(22.dp),
                        )
                        Celda("${serie.reps}${if (esDeTiempo) "s" else ""}", Modifier.weight(1f))
                        if (llevaPeso) {
                            Celda(peso(serie.weight), Modifier.weight(1f))
                        }
                        if (muestraRIR) {
                            Celda(serie.rir?.toString(), Modifier.weight(1f))
                        }
                    }
                }
            }
        }
    }
}

/** Las series prescritas en una sola forma, vengan parejas o detalladas. */
private fun series(item: RoutineExercise): List<PlannedSet> {
    val detalladas = item.sets
    if (!detalladas.isNullOrEmpty()) {
        return detalladas.mapIndexed { indice, serie -> serie.copy(setNumber = indice + 1) }
    }
    return (0 until maxOf(1, item.targetSets)).map {
        PlannedSet(
            setNumber = it + 1,
            weight = item.targetWeight,
            reps = item.targetReps,
            rir = item.targetRIR,
        )
    }
}

/** El resumen de lo prescrito: "3 × 10" cuando todas las reps son iguales, si
 *  no la lista separada por puntos. Con "s" al final si es ejercicio de
 *  tiempo. */
private fun resumen(series: List<PlannedSet>, esDeTiempo: Boolean): String {
    val reps = series.map { it.reps }
    val unidad = if (esDeTiempo) "s" else ""
    return if (reps.distinct().count() == 1) {
        "${reps.count()} × ${reps.first()}$unidad"
    } else {
        reps.joinToString(" · ") + unidad
    }
}

/** Un valor prescrito. Vacío se muestra como raya, no como cero. */
@Composable
private fun Celda(texto: String?, modifier: Modifier = Modifier) {
    val tema = LocalD2Theme.current
    Text(
        texto ?: "—",
        fontSize = 13.sp,
        color = if (texto == null) tema.texto3 else tema.texto,
        textAlign = TextAlign.Center,
        modifier = modifier
            .background(tema.vidrio(1), RoundedCornerShape(11.dp))
            .padding(vertical = 7.dp),
    )
}

/** El peso a "60 kg" o "60,5 kg": sin los ceros de relleno del Double. */
private fun peso(peso: Double?): String? = peso?.let {
    if (it % 1.0 == 0.0) {
        "${it.toLong()} kg"
    } else {
        "${String.format(Locale.US, "%.1f", it)} kg"
    }
}

File: ios/SiezaGym/Features/Routines/RoutineDetailScreen.swift
import SwiftUI

/// El detalle de una rutina, igual que `/rutinas/<id>` en la web: los tres
/// números, la lista de ejercicios que se despliega de a uno, el reparto
/// muscular, y abajo el botón de empezar.
struct RoutineDetailScreen: View {
    @Environment(\.tema) private var tema
    let routine: Routine
    let store: GymStore
    let onStart: (Routine) -> Void

    @State private var abierto: String?

    private var minutos: Int { RoutineSummary.estimatedMinutes(routine, catalog: store.catalog) }
    private var reparto: [RoutineSummary.MuscleShare] {
        RoutineSummary.muscleDistribution(routine, catalog: store.catalog)
    }
    private var hayGifs: Bool {
        routine.exercises.contains { store.exercise($0.exerciseID)?.mediaURL != nil }
    }

    var body: some View {
        ZStack(alignment: .bottom) {
            Pantalla(titulo: routine.name, rotulo: routine.isAssigned ? "Rutina del coach" : nil, volver: true) {
                StatsCard(datos: [
                    ("\(routine.exercises.count)", routine.exercises.count == 1 ? "ejercicio" : "ejercicios"),
                    ("\(routine.totalSets)", routine.totalSets == 1 ? "serie" : "series"),
                    // Es una cuenta sobre las series prescritas, no un tiempo
                    // medido: la pantalla lo dice.
                    ("\(minutos)", "min estimados"),
                ])
                .padding(.top, 20)
                .overlay(alignment: .top) {
                    Capsule()
                        .fill(tema.solido)
                        .frame(height: 4)
                        .padding(.horizontal, 28)
                        .padding(.top, 21)
                }

                SectionLabel("Ejercicios · \(routine.exercises.count)")
                    .padding(.top, 24)
                    .padding(.bottom, 10)

                if routine.exercises.isEmpty {
                    Vacio(texto: "Esta rutina no tiene ejercicios.")
                } else {
                    PanelLista {
                        ForEach(Array(routine.exercises.enumerated()), id: \.offset) { indice, item in
                            if indice > 0 {
                                Rectangle().fill(tema.borde).frame(height: 1)
                            }
                            FilaEjercicio(
                                item: item,
                                ejercicio: store.exercise(item.exerciseID),
                                nombre: store.name(of: item.exerciseID),
                                abierto: abierto == clave(indice, item),
                                alTocar: {
                                    abierto = abierto == clave(indice, item) ? nil : clave(indice, item)
                                }
                            )
                        }
                    }
                }

                if !reparto.isEmpty {
                    SectionLabel("Músculos que trabaja")
                        .padding(.top, 24)
                        .padding(.bottom, 10)

                    GlassCard(padding: 16) {
                        VStack(spacing: 11) {
                            ForEach(reparto.prefix(5)) { fila in
                                VStack(spacing: 5) {
                                    HStack {
                                        Text(fila.muscle.label)
                                            .font(.system(size: 12))
                                            .foregroundStyle(tema.texto)
                                        Spacer()
                                        Text("\(Int((fila.pct * 100).rounded()))%")
                                            .font(.system(size: 12))
                                            .foregroundStyle(tema.solido)
                                    }
                                    WidgetMeter(value: fila.pct)
                                }
                            }
                        }
                    }
                }

                if hayGifs { CreditoGifs() }

                // Aire para que el último ejercicio no quede abajo del botón.
                Color.clear.frame(height: 90)
            }

            boton
        }
        .bottomNavInset()
    }

    private func clave(_ indice: Int, _ item: RoutineExercise) -> String {
        "\(indice)-\(item.exerciseID)"
    }

    private var boton: some View {
        Button { onStart(routine) } label: {
            Label("Comenzar entrenamiento", systemImage: "play.fill")
        }
        .buttonStyle(SolidButtonStyle())
        .disabled(routine.exercises.isEmpty)
        .padding(.horizontal, 18)
        .padding(.bottom, 12)
    }
}

/// Un ejercicio de la rutina. Cerrado muestra el resumen de lo prescrito;
/// abierto, una fila por serie.
private struct FilaEjercicio: View {
    @Environment(\.tema) private var tema
    let item: RoutineExercise
    let ejercicio: Exercise?
    let nombre: String
    let abierto: Bool
    let alTocar: () -> Void

    /// Las series prescritas en una sola forma, vengan parejas o detalladas.
    private var series: [(numero: Int, reps: Int, peso: Double?, rir: Int?)] {
        if let detalladas = item.sets, !detalladas.isEmpty {
            return detalladas.enumerated().map { (indice, serie) in (indice + 1, serie.reps, serie.weight, serie.rir) }
        }
        return (0..<max(1, item.targetSets)).map {
            ($0 + 1, item.targetReps, item.targetWeight, item.targetRIR)
        }
    }

    private var esDeTiempo: Bool { ejercicio?.registrationType.isTimeBased == true }
    private var llevaPeso: Bool { ejercicio?.registrationType == .pesoReps }
    private var muestraRIR: Bool { series.contains { $0.rir != nil } }

    private var resumen: String {
        let reps = series.map(\.reps)
        let unidad = esDeTiempo ? "s" : ""
        if Set(reps).count == 1 { return "\(reps.count) × \(reps[0])\(unidad)" }
        return reps.map(String.init).joined(separator: " · ") + unidad
    }

    var body: some View {
        VStack(spacing: 0) {
            Button(action: alTocar) {
                HStack(spacing: 12) {
                    Miniatura(url: ejercicio?.mediaURL, lado: 54)
                    VStack(alignment: .leading, spacing: 4) {
                        Text(nombre)
                            .font(.system(size: 14, weight: .medium))
                            .foregroundStyle(tema.texto)
                            .lineLimit(1)
                        Text(ejercicio?.primaryMuscle?.label ?? "Sin datos")
                            .font(.system(size: 11))
                            .foregroundStyle(tema.texto2)
                    }
                    .frame(maxWidth: .infinity, alignment: .leading)

                    Text(resumen)
                        .font(.system(size: 12))
                        .foregroundStyle(tema.texto2)
                        .lineLimit(1)

                    Image(systemName: "chevron.down")
                        .font(.system(size: 12, weight: .medium))
                        .foregroundStyle(tema.texto3)
                        .rotationEffect(.degrees(abierto ? 180 : 0))
                }
                .padding(.horizontal, 14)
                .padding(.vertical, 12)
                .contentShape(.rect)
            }
            .buttonStyle(.plain)

            if abierto {
                VStack(spacing: 6) {
                    HStack(spacing: 8) {
                        Text("").frame(width: 22)
                        Text(esDeTiempo ? "Tiempo" : "Reps").frame(maxWidth: .infinity)
                        if llevaPeso { Text("Peso").frame(maxWidth: .infinity) }
                        if muestraRIR { Text("RIR").frame(maxWidth: .infinity) }
                    }
                    .font(.system(size: 10))
                    .foregroundStyle(tema.texto3)

                    ForEach(series, id: \.numero) { serie in
                        HStack(spacing: 8) {
                            Text("\(serie.numero)")
                                .font(.system(size: 11))
                                .foregroundStyle(tema.texto3)
                                .frame(width: 22, alignment: .leading)
                            celda("\(serie.reps)\(esDeTiempo ? "s" : "")")
                            if llevaPeso {
                                celda(serie.peso.map { "\($0.formatted()) kg" })
                            }
                            if muestraRIR {
                                celda(serie.rir.map(String.init))
                            }
                        }
                    }
                }
                .padding(.horizontal, 14)
                .padding(.bottom, 14)
            }
        }
        .background(abierto ? tema.solido.opacity(0.08) : .clear)
        .overlay(alignment: .leading) {
            if abierto {
                RoundedRectangle(cornerRadius: 3)
                    .fill(tema.solido)
                    .frame(width: 3)
                    .padding(.vertical, 10)
            }
        }
        .animation(.snappy(duration: 0.2), value: abierto)
    }

    /// Un valor prescrito. Vacío se muestra como raya, no como cero.
    private func celda(_ texto: String?) -> some View {
        Text(texto ?? "—")
            .font(.system(size: 13))
            .foregroundStyle(texto == nil ? tema.texto3 : tema.solido)
            .frame(maxWidth: .infinity, minHeight: 32)
            .background(texto == nil ? tema.vidrio(1) : tema.solido.opacity(0.08), in: .rect(cornerRadius: 11))
    }
}
